### PR TITLE
Flash error improvement

### PR DESCRIPTION
**Pull request by Ioan-Cristian**
### Pull Request Overview

This pull request addresses the issue #3692 .

### Testing Strategy

This pull request has not been tested on any hardware yet

### TODO or Help Wanted

I need testing on actual hardware to see if everything works as expected.

### Documentation Updated

No updates required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/extra/src/log.rs [hunk 1]
```diff
@@ -824,7 +824,7 @@ impl<'a, F: Flash + 'static> flash::Client<F> for Log<'a, F> {
                     _ => unreachable!(),
                 }
             }
-            flash::Error::FlashError | flash::Error::FlashMemoryProtectionError => {
+            Err(_) => {
```
Comment (bradjc) [inline capsules/extra/src/log.rs:827]: It would be nice to not change that these are specified (and future errors would have to be added explicitly).

File: capsules/extra/src/log.rs [hunk 2]
```diff
@@ -872,7 +872,7 @@ impl<'a, F: Flash + 'static> flash::Client<F> for Log<'a, F> {
                     }
                 }
             }
-            flash::Error::FlashError | flash::Error::FlashMemoryProtectionError => {
+            Err(_) => {
```
Comment (bradjc) [inline capsules/extra/src/log.rs:875]: Here as well.

File: capsules/extra/src/log.rs (hunk 1 above)
Comment (Ioan-Cristian) [reply capsules/extra/src/log.rs:827]: It looks like the code inside the block doesn't care if it is `FlashError`, `FlashMemoryProtectionError`, nor any other error. It matters only if it's an error or not. Explicitly matching each error code might just annoy the programmer without any benefit for the code's logic. Am I wrong?

Comment (bradjc) [reply capsules/extra/src/log.rs:827]: Well I'm not sure that the original author didn't specifically choose to not use the catchall. It's easier for me to approve this PR if the change is just mechanistic, and not actually making different design decisions.

Comment (Ioan-Cristian) [reply capsules/extra/src/log.rs]: Matched errors explicitly.

File: capsules/extra/src/log.rs (hunk 2 above)
Comment (Ioan-Cristian) [reply capsules/extra/src/log.rs:879]: Matched errors explicitly.

Comment (bradjc) [reply capsules/extra/src/log.rs:879]: Sometimes github hides things, did this one get updated?

Comment (Ioan-Cristian) [reply capsules/extra/src/log.rs]: No, it didn't. I apologize. Now, it has been updated.

**Review by bradjc — APPROVED** [review]

**Comment by twilfredo** [comment]
might be a good idea to rebase and squish the last 3 commits into a single one. Looks like there's some overlap + a merge commit?

**Comment by Ioan-Cristian** [comment]
> might be a good idea to rebase and squish the last 3 commits into a single one. Looks like there's some overlap + a merge commit?

Done.

**Comment by Ioan-Cristian** [comment]
Rebased.

**Review by bradjc — APPROVED** [review]

File: boards/opentitan/src/tests/flash.rs [hunk 3]
```diff
@@ -2,474 +2,417 @@
...
-            let key_input = static_init!(
-                [u8; 16],
-                [
-                    0x12, 0x34, 0x56, 0x78, 0x9A, 0xBC, 0xDE, 0xF0, 0x12, 0x34, 0x56, 0x78, 0x9A,
-                    0xBC, 0xDE, 0xF0
-                ]
-            );
-            let key = static_init!([u8; 8], [0; 8]);
-            let value = static_init!([u8; 3], [0x10, 0x20, 0x30]);
-            let ret = static_init!([u8; 4], [0; 4]);
```
Comment (alistair23) [inline boards/opentitan/src/tests/flash.rs:36]: Why did you remove test cases? Does this PR cause a regression but instead of fixing it you removed the tests?

Comment (Ioan-Cristian) [reply boards/opentitan/src/tests/flash.rs:36]: Pretty sure I didn't remove them. @bradjc may confirm this.

I didn't perform any tests on actual hardware so I could not be aware of any regression tests. Initially, I only changed the signature of flash methods and a small detail design implementation. However, the detail design was rejected by @bradjc , so there was not any change in code logic.

The sole explanation that I have: it looks like my pull request was merged on 10th October, while yours on 9th October, so your work was overwritten by my pull request. Locally, I still have `tick_test.rs` file that you removed. I apologize for my mistake!

I can fix this, but I would require someone to run the test suite on OpenTitan. What do you think?

Comment (alistair23) [reply boards/opentitan/src/tests/flash.rs:36]: > The sole explanation that I have: it looks like my pull request was merged on 10th October, while yours on 9th October, so your work was overwritten by my pull request. Locally, I still have `tickv_test.rs` file that you removed. I apologize for my mistake!

Urgh. I would hope that would have been caught by review to ensure we don't get nasty merge conflicts like this.

That's ok! I was just confused why test cases are being removed
